// @hediet/json-rpc/dist/Channel.js
import { constValue } from "./common";
/**
 * Once a channel is constructed, it processes all incoming messages.
 */
export class Channel {
    connect;
    constructor(connect) {
        this.connect = connect;
    }
    mapContext(map) {
        return new Channel((listener) => this.connect(listener ? mapRequestHandlerContext(listener, map) : undefined));
    }
}
/**
 * Implements a channel that directly forwards the requests to the given request handler.
 */
export class LoopbackChannel {
    requestHandler;
    id = 0;
    constructor(requestHandler) {
        this.requestHandler = requestHandler;
    }
    sendRequest(request, context, messageIdCallback) {
        const curId = this.id++;
        if (messageIdCallback) {
            messageIdCallback(curId);
        }
        return this.requestHandler.handleRequest(request, curId, context);
    }
    sendNotification(notification, context) {
        return this.requestHandler.handleNotification(notification, context);
    }
    toString() {
        return "Loopback";
    }
    state = constValue({ state: "open" });
}
export function mapMessageSenderContext(messageSender, map) {
    return {
        sendNotification: (notification, context) => messageSender.sendNotification(notification, map(context)),
        sendRequest: (request, context, messageIdCallback) => messageSender.sendRequest(request, map(context), messageIdCallback),
        toString: () => messageSender.toString(),
        state: messageSender.state,
    };
}
export function mapRequestHandlerContext(messageHandler, map) {
    return {
        handleNotification: (request, context) => messageHandler.handleNotification(request, map(context)),
        handleRequest: (request, requestId, context) => messageHandler.handleRequest(request, requestId, map(context)),
    };
}
