// @hediet/json-rpc/dist/common.js
export class EventEmitter {
    listeners = new Set();
    event = (listener) => {
        this.listeners.add(listener);
        return {
            dispose: () => {
                this.listeners.delete(listener);
            }
        };
    };
    fire(args) {
        this.listeners.forEach(listener => listener(args));
    }
}
export function constValue(value) {
    const eventEmitter = new EventEmitter();
    return {
        get value() {
            return value;
        },
        get onChange() {
            return eventEmitter.event;
        }
    };
}
export class ValueWithChangeEvent {
    _value;
    eventEmitter;
    constructor(initialValue) {
        this._value = initialValue;
        this.eventEmitter = new EventEmitter();
    }
    get value() {
        return this._value;
    }
    set value(newValue) {
        if (this._value !== newValue) {
            this._value = newValue;
            this.eventEmitter.fire(newValue);
        }
    }
    get onChange() {
        return this.eventEmitter.event;
    }
}
export function createTimeout(delay, callback) {
    const handle = setTimeout(callback, delay);
    return {
        dispose: () => clearTimeout(handle)
    };
}
export function setAndDeleteOnDispose(set, keyOrItem, item) {
    if (set instanceof Set) {
        set.add(keyOrItem);
        return { dispose: () => set.delete(keyOrItem) };
    }
    else {
        set.set(keyOrItem, item);
        return { dispose: () => set.delete(keyOrItem) };
    }
}
export class Deferred {
    _state = "none";
    promise;
    resolve = () => { };
    reject = () => { };
    constructor() {
        this.promise = new Promise((resolve, reject) => {
            this.resolve = resolve;
            this.reject = reject;
        });
    }
    get state() {
        return this._state;
    }
}
export class Barrier {
    deferred = new Deferred();
    unlock = this.deferred.resolve;
    reject = this.deferred.reject;
    onUnlocked = this.deferred.promise;
    constructor() { }
    get state() {
        return this.deferred.state;
    }
}
