// @hediet/json-rpc/dist/schema.js
export var Serializers;
(function (Serializers) {
    function sAny() {
        return {
            deserializeFromJson: input => ({ hasErrors: false, value: input }),
            serializeToJson: input => input,
        };
    }
    Serializers.sAny = sAny;
    function sEmptyObject() {
        return {
            deserializeFromJson: input => ({ hasErrors: false, value: {} }),
            serializeToJson: input => ({}),
        };
    }
    Serializers.sEmptyObject = sEmptyObject;
    function sVoidFromNull() {
        return {
            deserializeFromJson: input => ({ hasErrors: false, value: undefined }),
            serializeToJson: input => null,
        };
    }
    Serializers.sVoidFromNull = sVoidFromNull;
})(Serializers || (Serializers = {}));
let globalMapper = {
    serializer: val => {
        if (!isSerializer(val)) {
            return undefined;
        }
        return val;
    },
    zod: val => {
        if (!isZodType(val)) {
            return undefined;
        }
        return {
            serializeToJson: input => val.parse(input),
            deserializeFromJson: input => {
                const result = val.safeParse(input);
                if (result.success) {
                    return { hasErrors: false, value: result.data };
                }
                else {
                    return { hasErrors: true, value: undefined, errorMessage: result.error.message };
                }
            }
        };
    }
};
function isZodType(val) {
    return 'parse' in val && 'safeParse' in val;
}
function isSerializer(val) {
    return 'serializeToJson' in val && 'deserializeFromJson' in val;
}
export function setMapper(mapper) {
    Object.assign(globalMapper, mapper);
}
export function convertSerializer(val) {
    for (const [key, value] of Object.entries(globalMapper)) {
        const result = value(val);
        if (result) {
            return result;
        }
    }
    throw new Error(`No serializer found for value: ${val}`);
}
;
