// @hediet/json-rpc-websocket/dist/server.js
import * as WebSocket from "ws";
import { WebSocketTransport } from "./index";
import { EventEmitter, Barrier } from "@hediet/json-rpc";
/**
 * Starts a new web socket server.
 * Use `handleConnection` to handle the streams of incoming connections.
 */
export function startWebSocketServer(options, handleConnection) {
    const opts = {};
    const l = options.listenOn;
    if ("port" in l) {
        opts.port = l.port === "random" ? 0 : l.port;
    }
    else if ("server" in l) {
        opts.server = l.server;
    }
    const wss = new WebSocket.Server(opts);
    wss.on("connection", ws => {
        const stream = WebSocketTransport.fromWebSocket(ws);
        handleConnection(stream);
    });
    return new WebSocketServer(wss);
}
/**
 * Wraps a websocket server.
 */
export class WebSocketServer {
    server;
    constructor(server) {
        this.server = server;
        const listeningEventEmitter = new Barrier();
        this.onListening = listeningEventEmitter.onUnlocked;
        let state = "None";
        server.on("listening", () => {
            listeningEventEmitter.unlock();
            state = "Listening";
        });
        const errorEventEmitter = new EventEmitter();
        this.onError = errorEventEmitter.event;
        server.on("error", error => {
            if (state == "None") {
                state = "Error";
                listeningEventEmitter.reject(error);
            }
            errorEventEmitter.fire(error);
        });
    }
    /**
     * Is resolved once the socket is listening.
     */
    onListening;
    onError;
    /**
     * Gets the port the websocket server is running on.
     */
    get port() {
        const addrInfo = this.server.address();
        return addrInfo.port;
    }
    /**
     * Closes the server.
     */
    close() {
        this.server.close();
    }
    /**
     * Same as `close`.
     */
    dispose() {
        this.close();
    }
}
